linker-plugin LTO doesn't work with --wrap

The package itself requires a couple of dependencies: `xorg-util-macros font-util meson xtrans` and reproduction steps are the following:

```
wget https://www.x.org/pub/individual/xserver/xorg-server-21.1.22.tar.xz
tar xvJf xorg-server-21.1.22.tar.xz
cd xorg*
mkdir build
cd build
CFLAGS="-B /home/marxin/Programming/wild" meson setup -Db_lto=true ..
ninja
...
cc  -o test/tests test/tests.p/.._mi_miinitext.c.o test/tests.p/.._mi_micmap.c.o test/tests.p/fixes.c.o test/tests.p/input.c.o test/tests.p/list.c.o test/tests.p/misc.c.o test/tests.p/signal-logging.c.o test/tests.p/string.c.o test/tests.p/test_xkb.c.o test/tests.p/tests-common.c.o test/tests.p/tests.c.o test/tests.p/touch.c.o test/tests.p/xfree86.c.o test/tests.p/xtest.c.o test/tests.p/hashtabletest.c.o test/tests.p/xi1_protocol-xchangedevicecontrol.c.o test/tests.p/xi2_protocol-common.c.o test/tests.p/xi2_protocol-xiqueryversion.c.o test/tests.p/xi2_protocol-xiquerydevice.c.o test/tests.p/xi2_protocol-xiselectevents.c.o test/tests.p/xi2_protocol-xigetselectedevents.c.o test/tests.p/xi2_protocol-xisetclientpointer.c.o test/tests.p/xi2_protocol-xigetclientpointer.c.o test/tests.p/xi2_protocol-xiquerypointer.c.o test/tests.p/xi2_protocol-xipassivegrabdevice.c.o test/tests.p/xi2_protocol-xiwarppointer.c.o test/tests.p/xi2_protocol-eventconvert.c.o test/tests.p/xi2_xi2.c.o -flto=auto -Wl,--as-needed -Wl,--no-undefined -fvisibility=hidden -B /home/marxin/Programming/wild -Wl,--start-group mi/liblibxserver_mi.a dix/liblibxserver_dix.a composite/liblibxserver_composite.a damageext/liblibxserver_damageext.a dbe/liblibxserver_dbe.a randr/liblibxserver_randr.a miext/damage/liblibxserver_miext_damage.a render/liblibxserver_render.a present/liblibxserver_present.a Xext/liblibxserver_xext.a miext/sync/liblibxserver_miext_sync.a xfixes/liblibxserver_xfixes.a Xi/liblibxserver_xi.a xkb/liblibxserver_xkb.a record/liblibxserver_record.a os/liblibxserver_os.a os/liblibxlibc.a dri3/liblibxserver_dri3.a glx/liblibglxvnd.a hw/xfree86/common/libxorg_common.a hw/xfree86/loader/libxorg_loader.a hw/xfree86/ddc/libxorg_ddc.a hw/xfree86/xkb/libxorg_xkb.a hw/xfree86/i2c/libxorg_i2c.a hw/xfree86/modes/libxorg_modes.a hw/xfree86/os-support/libxorg_os_support.a hw/xfree86/parser/libxorg_parser.a hw/xfree86/ramdac/libxorg_ramdac.a fb/liblibxserver_fb.a Xext/liblibxserver_xext_vidmode.a dix/liblibxserver_main.a config/liblibxserver_config.a hw/xfree86/dri/libxorg_dri.a hw/xfree86/dri2/libxorg_dri2.a -Wl,-wrap,dixLookupWindow -Wl,-wrap,dixLookupClient -Wl,-wrap,WriteToClient -Wl,-wrap,dixLookupWindow -Wl,-wrap,XISetEventMask -Wl,-wrap,AddResource -Wl,-wrap,GrabButton /usr/lib/libpixman-1.so /usr/lib/libxcvt.so /usr/lib/libsystemd.so /usr/lib/libbsd.so /usr/lib/libxkbfile.so /usr/lib/libXfont2.so /usr/lib/libXdmcp.so -lm /usr/lib/libxshmfence.so -ldl -lmd /usr/lib/libtirpc.so -lpthread /usr/lib/libXau.so /usr/lib/libdrm.so /usr/lib/libGL.so /usr/lib/libdbus-1.so /usr/lib/libpciaccess.so /usr/lib/libudev.so -Wl,--end-group
wild: error: Failed copying from test/tests.p/xi2_protocol-xiselectevents.c.o (3584 (14/0)) to output file
  Caused by:
    Failed to apply relocations in section `.text` of test/tests.p/xi2_protocol-xiselectevents.c.o
    Failed to apply relocation of type R_X86_64_PLT32 to symbol `__real_XISetEventMask` (1583 local=35) in file #3584 (14/0) (test/tests.p/xi2_protocol-xiselectevents.c.o) defined as 22226 in file #17040 (66/144) (LTO input `Xi/liblibxserver_xi.a.p/exevents.c.o`) (ABSOLUTE) at offset 0x281
    Missing resolution for: symbol `__real_XISetEventMask` (1583 local=35) in file #3584 (14/0) (test/tests.p/xi2_protocol-xiselectevents.c.o) defined as 22226 in file #17040 (66/144) (LTO input `Xi/liblibxserver_xi.a.p/exevents.c.o`) (ABSOLUTE)

collect2: error: ld returned 255 exit status
ninja: build stopped: subcommand failed.
```

--- a/libwild/src/lib.rs
+++ b/libwild/src/lib.rs
@@ -303,8 +303,6 @@ impl Linker {
             loaded,
         )?;
 
-        // TODO: Doing this here means that we can't wrap symbols produced by the linker plugin.
-        // Moving it earlier or later however requires some rethought as to how this works.
         symbol_db.apply_wrapped_symbol_overrides();
 
         let mut resolver = resolution::Resolver::default();

--- a/libwild/src/linker_plugins.rs
+++ b/libwild/src/linker_plugins.rs
@@ -25,6 +25,7 @@ use crate::input_data::FileLoader;
 use crate::input_data::InputRef;
 use crate::layout_rules::LayoutRulesBuilder;
 use crate::output_section_id::OutputSections;
+use crate::platform::Args as _;
 use crate::platform::Platform;
 use crate::platform::RawSymbolName as _;
 use crate::resolution::ResolutionResources;
@@ -211,6 +212,11 @@ impl<'data> LinkerPlugin<'data> {
 
         timing_phase!("Linker plugin codegen");
 
+        // Mark wrapped symbol names and their __wrap_/__real_ variants as referenced by non-IR
+        // code. This ensures the plugin keeps them in the LTO output rather than
+        // internalising/removing them.
+        mark_wrap_symbols_as_non_ir_ref(symbol_db, per_symbol_flags);
+
         let plugin_outputs = self.store.loaded()?.with_callbacks(|callbacks| {
             if let Some(cb) = callbacks.all_symbols_read {
                 let ctx = AllSymbolsReadContext {
@@ -237,6 +243,10 @@ impl<'data> LinkerPlugin<'data> {
             layout_rules_builder,
             plugin_loaded,
         )?;
+
+        // Re-apply --wrap overrides so that wrapped names now map to definitions from the LTO
+        // output rather than the LTO input objects.
+        symbol_db.apply_wrapped_symbol_overrides();
 
         resolver.resolve_symbols_and_select_archive_entries(symbol_db, per_symbol_flags)?;
 
@@ -922,6 +932,20 @@ fn get_symbol_resolution<'data>(
     if !sym.version.is_null() {
         raw_name.version_name = Some(unsafe { CStr::from_ptr(sym.version) }.to_bytes());
     }
+
+    // If the symbol was wrapped via --wrap, the name table has been modified to map the original
+    // name to __wrap_<name>. We must not report the wrapped resolution to the plugin, since the
+    // plugin needs to know the pre-wrap state.
+    let wrap_names = symbol_db.args.symbol_names_to_wrap();
+    let is_wrapped = wrap_names.iter().any(|w| w.as_bytes() == raw_name.name);
+    if is_wrapped {
+        return if sym.is_undefined() {
+            PluginSymbolResolution::ResolvedExec
+        } else {
+            PluginSymbolResolution::PreemptedReg
+        };
+    }
+
     let symbol_id = symbol_db
         .get(&PreHashedSymbolName::from_raw(&raw_name), true)
         .map(|id| symbol_db.definition(id));
@@ -1417,4 +1441,25 @@ pub(crate) fn resolve_lto_symbols<'data, 'scope>(
                 Ok(())
             },
         )
+}
+
+/// Marks symbols related to --wrap as having non-IR references. This ensures that the linker
+/// plugin preserves these symbols in its output rather than internalising them.
+fn mark_wrap_symbols_as_non_ir_ref<'data, P: Platform>(
+    symbol_db: &SymbolDb<'data, P>,
+    per_symbol_flags: &mut PerSymbolFlags,
+) {
+    for name in symbol_db.args.symbol_names_to_wrap() {
+        for lookup_name in [
+            name.clone(),
+            format!("__wrap_{name}"),
+            format!("__real_{name}"),
+        ] {
+            if let Some(symbol_id) =
+                symbol_db.get_unversioned(&UnversionedSymbolName::prehashed(lookup_name.as_bytes()))
+            {
+                per_symbol_flags.set_flag(symbol_id, ValueFlags::HAS_NON_IR_REF);
+            }
+        }
+    }
 }

--- a/libwild/src/symbol_db.rs
+++ b/libwild/src/symbol_db.rs
@@ -586,7 +586,16 @@ impl<'data, P: Platform> SymbolDb<'data, P> {
 
         for name in wrap {
             let name_bytes = allocator.alloc_slice_copy(name.as_bytes());
-            let orig_id = self.get_unversioned(&UnversionedSymbolName::prehashed(name_bytes));
+            let real_name = allocator.alloc_slice_copy(format!("__real_{name}").as_bytes());
+
+            // When this function is called a second time (after LTO), the name table already has
+            // "foo" mapped to __wrap_foo's symbol ID from the first call. To get the ORIGINAL foo
+            // symbol ID, we first check if __real_foo already has a mapping (set by the first
+            // call), and fall back to looking up "foo" only on the first call.
+            let orig_id = self
+                .get_unversioned(&UnversionedSymbolName::prehashed(real_name))
+                .or_else(|| self.get_unversioned(&UnversionedSymbolName::prehashed(name_bytes)));
+
             let wrap_name = format!("__wrap_{name}");
             if let Some(wrap_id) =
                 self.get_unversioned(&UnversionedSymbolName::prehashed(wrap_name.as_bytes()))
@@ -595,7 +604,6 @@ impl<'data, P: Platform> SymbolDb<'data, P> {
             }
 
             if let Some(orig_id) = orig_id {
-                let real_name = allocator.alloc_slice_copy(format!("__real_{name}").as_bytes());
                 self.override_name(UnversionedSymbolName::prehashed(real_name), orig_id);
             }
         }

--- a/wild/tests/external_tests/mold_skip_tests.toml
+++ b/wild/tests/external_tests/mold_skip_tests.toml
@@ -183,7 +183,6 @@ tests = [
   "lto-llvm2.sh",           # -m llvm is not yet supported
   "duplicate-error-lto.sh", # Wild deduplicates inputs
   "lto-comdat.sh",          # We don't support COMDAT
-  "wrap-lto.sh",            # We don't yet support --wrap with LTO
   "lto-archive4.sh",
   "lto-mixed-gcc-fat.sh",
 ]

--- a/wild/tests/sources/elf/wrap-lto/wrap-lto-2.c
+++ b/wild/tests/sources/elf/wrap-lto/wrap-lto-2.c
@@ -1,0 +1,1 @@
+int foo(void) { return 10; }

--- a/wild/tests/sources/elf/wrap-lto/wrap-lto.c
+++ b/wild/tests/sources/elf/wrap-lto/wrap-lto.c
@@ -1,0 +1,30 @@
+//#AbstractConfig:default
+//#RequiresLinkerPlugin:true
+//#Object:runtime.c
+//#Object:wrap-lto-2.c:-fno-lto
+//#SkipLinker:ld
+//#CompArgs:-flto
+//#LinkArgs:-flto -nostdlib -z now -Wl,-wrap,foo
+
+//#Config:gcc:default
+//#LinkerDriver:gcc
+
+//#Config:clang:default
+//#Compiler:clang
+//#LinkerDriver:clang
+//#EnableLinker:lld
+
+#include "../common/runtime.h"
+
+int foo(void);
+int __real_foo(void);
+
+int __wrap_foo(void) { return __real_foo() + 32; }
+
+void _start(void) {
+  runtime_init();
+  if (foo() != 42) {
+    exit_syscall(100);
+  }
+  exit_syscall(42);
+}

